fdbserver/LogSystemPeekCursor.actor.cpp

```diff
@@ -1019,13 +1019,26 @@ ACTOR Future<Void> bufferedGetMore( ILogSystem::BufferedCursor* self, TaskPriori
 	} else {
 		uniquify(self->messages);
 	}
-	self->messageIndex = 0;
-	self->hasNextMessage = self->messages.size() > 0;
 	Version minVersion = self->end;
 	for(auto& cursor : self->cursors) {
 		minVersion = std::min(minVersion, cursor->version().version);
 	}
-	self->messageVersion = LogMessageVersion(minVersion);
+	self->messageVersion = LogMessageVersion(std::max(minVersion, self->poppedVersion));
+	// If the tags have been popped inconsistently, then we might have messages we need to throw away.
+	if (self->poppedVersion <= minVersion) {
+		self->hasNextMessage = self->messages.size() > 0;
+		self->messageIndex = 0;
+	} else {
+	  if (self->messages.size() > 0 && self->messages[self->messages.size()-1].version < self->messageVersion) {
+			self->hasNextMessage = false;
+			self->messageIndex = 0;
+		} else {
+			auto iter = std::lower_bound(self->messages.begin(), self->messages.end(),
+					ILogSystem::BufferedCursor::BufferedMessage(self->arena(), LiteralStringRef(""), {}, self->messageVersion));
+			self->hasNextMessage = iter != self->messages.end();
+			self->messageIndex = self->hasNextMessage ? iter - self->messages.begin() : 0;
```

Comment:
This all is actually an implementation of advance_to.

lower_bound is a binary search when given a random access iterator